firmware/src/hal/sdl/hal_sdl.cpp

```
@@ -9,62 +11,245 @@
 
+int read_input(InputEvent* out, int capacity) {
+  const int count = input_count_ < capacity ? input_count_ : capacity;
+  for (int i = 0; i < count; ++i) out[i] = input_[i];
+  for (int i = count; i < input_count_; ++i) input_[i - count] = input_[i];
+  input_count_ -= count;
+  return count;
+}
+
+void start_timer(uint32_t period_us, TimerCallback callback) {
+  timer_callback_ = callback;
+  if (timer_ != 0) return;  // already ticking: the new callback is all that changes
```

Clang (14.0.6)

[warning] statement should be inside braces

(readability-braces-around-statements)